### ISSUE TITLE
LDAPException(resultCode=84 (decoding error)

### ISSUE THREAD

**Issue by nikolai-krasilnikov** [pingidentity/ldapsdk#106]
Hi, I am trying to connect to an LDAP server from simple Adnroid app where i have login form (AD username and password), but I get the following error:

**LDAPException(resultCode=84 (decoding error), errorMessage='The connection to server xxx.x.xxx.xx:443 was closed while waiting for a response to a bind request SimpleBindRequest():  Unable to read or decode an LDAP message:  Invalid value length of 80 for an ASN.1 integer element.  Integer element values must have a length between 1 and 4 bytes., ldapSDKVersion=5.1.4, revision=95121137bbf71bc02638a53c2c8ddf3588c2bcc4')**

An error occurs when I try to log in. Can I log in using only uid in bindDN? I tried specifying the full DN (dn="CN=LastName FirstName,OU=Users,OU=xx,OU=xxx,OU=xxx,DC=xxx,DC=xxx,DC=xxx,DC=xxx,DC=su") and (dn="uid=xxx,OU=Users,OU=xx,OU=xxx,OU=xxx,DC=xxx,DC=xxx,DC=xxx,DC=xxx,DC=su") , but got the same error.

What could be the problem? Thanks in advance.

Here is my code:

public class MainActivity extends Activity implements OnClickListener {
  
    int port=443;
    final String dn ="uid=xxx";
    final String password ="xxx";
    final String hostname = "xxx.x.xxx.xx";
    boolean login_flag=true;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        Button bt_login=(Button)findViewById(R.id.bt_login);
        bt_login.setOnClickListener(this);
    }

    @SuppressLint("SimpleDateFormat")
    @Override
    public void onClick(View view)
    {       
        new Thread(new Runnable()
        {
            @Override
            public void run()
            {
                try {
                    final BindRequest bindRequest = new SimpleBindRequest(dn,password);
                    final SSLUtil sslUtil = new SSLUtil(new TrustAllTrustManager());
                    final LDAPConnectionOptions connectionOptions = new LDAPConnectionOptions();
                    connectionOptions.setFollowReferrals(true);
                    final LDAPConnection ldapConnection = new LDAPConnection(sslUtil.createSSLSocketFactory(),hostname,port);
                    final BindResult bindResult = ldapConnection.bind(bindRequest);
                    final ResultCode resultCode = bindResult.getResultCode();
                    if(resultCode.equals(ResultCode.SUCCESS))
                    {
                        System.out.println("success");
                    }
                } catch (LDAPException e) {
                    login_flag=false;
                    e.printStackTrace();
                    System.out.println("No connection was established");
                } catch(Exception e) {
                    e.printStackTrace();
                } finally{
                    if(login_flag){
                        ldapConnection.close();
                        System.out.println("Connection Closed successfully");

                    }
                }
            }
        }).start();
    }

}

**Comment by dirmgr** [comment]
It looks like the problem is that you're sending an LDAPS request to port 443, which is typically used for HTTPS. If that's the case, then you're probably sending an LDAPS request to an HTTPS server, which the HTTPS server can't understand. In turn, it's returning an HTTPS error, which the LDAPS client can't understand.

The default standard port for LDAPS is 636, so I'd recommend trying that. If that doesn't work, then you should check with the server administrator to see what port you should use.

**Comment by nikolai-krasilnikov** [comment]
> It looks like the problem is that you're sending an LDAPS request to port 443, which is typically used for HTTPS. If that's the case, then you're probably sending an LDAPS request to an HTTPS server, which the HTTPS server can't understand. In turn, it's returning an HTTPS error, which the LDAPS client can't understand.
> 
> The default standard port for LDAPS is 636, so I'd recommend trying that. If that doesn't work, then you should check with the server administrator to see what port you should use.

Thanks a lot. I contacted the administrators of the LDAP server and it turned out that my collegue gave me incorrect connection data.

My task is to make authorization using the aCSPolicyName(Ldap-Display-Name) and user AD password (because CN is too long (example Ivanov Ivan Ivanovich, but aCSPolicyName looks like - ivanov121212) and it is easier for the user to remember one login and password for all systems) which is used by the user to log into Windows, but in the android application. I need to send aCSPolicyName and user password to check authorization.

As far as I understand, I can do this by searching in AD using a service account filtered by aCSPolicyName. Having found the user by taking his DN, and check the authorization.

Or can everything be made simpler? Can you give me some advice or an example. Thanks in advance.

**Comment by dirmgr** [comment]
Yes, that's generally how things work when using an LDAP simple bind: you first perform a search to find the entry for the target user and get their DN, and then you bind with that user's DN and password. The LDAP does provide support for SASL authentication, which is a framework that allows for other possibilities, and there are mechanisms that allow you to specify a username rather than a DN, but I'm pretty sure that Active Directory doesn't support most of them, and the couple of them that it does support (like GSSAPI) aren't available for use on Android.

Whether or not you need to bind first using a service account so that you can perform the search to get the target user's DN depends on the configuration for the server that you're using. It may be configured to allow that initial search to be performed anonymously, but clearly there is a security risk associated with that, and it's something that should only be considered on a private network and not the public internet (so for an Android app, it would really only be an option if you're using WiFi rather than the mobile provider's network). If anonymous searches aren't allowed, then you're kind of stuck because you probably don't want to hard-code the credentials for the service account into your app.

Although it's not a great alternative, if your app is only going to be distributed to a restricted audience (for example, employees of your company), then one possibility would be to use a service account with very limited access (basically, only allowed to obtain the DN of an entry for a given username), and to have the DN of that account hard-coded in the app, but then require the user to provide the password for that account the first time they use the app.

Unfortunately, without support for a better authentication mechanism like SASL PLAIN (which is like LDAP simple authentication but allows you to provide a username instead of a DN), there aren't many great, standards-compliant options available.

**Comment by nikolai-krasilnikov** [comment]
> Yes, that's generally how things work when using an LDAP simple bind: you first perform a search to find the entry for the target user and get their DN, and then you bind with that user's DN and password. The LDAP does provide support for SASL authentication, which is a framework that allows for other possibilities, and there are mechanisms that allow you to specify a username rather than a DN, but I'm pretty sure that Active Directory doesn't support most of them, and the couple of them that it does support (like GSSAPI) aren't available for use on Android.
> 
> Whether or not you need to bind first using a service account so that you can perform the search to get the target user's DN depends on the configuration for the server that you're using. It may be configured to allow that initial search to be performed anonymously, but clearly there is a security risk associated with that, and it's something that should only be considered on a private network and not the public internet (so for an Android app, it would really only be an option if you're using WiFi rather than the mobile provider's network). If anonymous searches aren't allowed, then you're kind of stuck because you probably don't want to hard-code the credentials for the service account into your app.
> 
> Although it's not a great alternative, if your app is only going to be distributed to a restricted audience (for example, employees of your company), then one possibility would be to use a service account with very limited access (basically, only allowed to obtain the DN of an entry for a given username), and to have the DN of that account hard-coded in the app, but then require the user to provide the password for that account the first time they use the app.
> 
> Unfortunately, without support for a better authentication mechanism like SASL PLAIN (which is like LDAP simple authentication but allows you to provide a username instead of a DN), there aren't many great, standards-compliant options available.


Hello, and thanks for your responses.

What am I doing wrong when trying to check if a user with attribute "sAMAccountName", "nikolai8332" in a specific AD group of the form "CN = xxx, OU = xxx, OU = xx, OU = xxx, OU = Groups, DC = xx, DC = xxxx, DC = xxx, DC = xxxx, DC = xx ". If I do not specify the DN of the AD group, in only "DC = xx, DC = xxxx, DC = xxx, DC = xxxx, DC = xx" then everything works and the code returns the attributes I need.

public void onClick(View view)
    {
        
        new Thread(new Runnable()
        {
            @Override
            public void run()
            {
                
                try {
                   

                    final BindRequest bindRequest = new SimpleBindRequest(dn,password);
                    final SSLUtil sslUtil = new SSLUtil(new TrustAllTrustManager());
                    final LDAPConnectionOptions connectionOptions = new LDAPConnectionOptions();
                    connectionOptions.setFollowReferrals(true);
                    final LDAPConnection ldapConnection = new LDAPConnection(sslUtil.createSSLSocketFactory(),hostname,port);

                    System.out.println("con");
                    final BindResult bindResult = ldapConnection.bind(bindRequest);
                    System.out.println("bind");

                    ResultCode resultCode = bindResult.getResultCode();
                    System.out.println("res");
                    if(resultCode.equals(ResultCode.SUCCESS))
                    {
                        System.out.println("success");
                    }

                    Filter filter = Filter.createEqualityFilter("sAMAccountName", "nikolai8332");
                    //Filter filter1 = Filter.createEqualityFilter()
                    SearchRequest searchRequest =
                            new SearchRequest("CN=xxx,OU=xxx,OU=xx,OU=xxx,OU=Groups,DC=xx,DC=xxxx,DC=xxx,DC=xxxx,DC=xx", 
SearchScope.SUB, filter,
                                    "distinguishedName","cn", "mail");
                    SearchResult searchResult;
                    System.out.println("start_search");
                    try
                    {
                        searchResult = ldapConnection.search(searchRequest);
                        System.out.println("search");
                        System.out.println(searchResult);

                        for (SearchResultEntry entry : searchResult.getSearchEntries())
                        {
                            String dn = entry.getDN();
                            String name = entry.getAttributeValue("cn");
                            String mail = entry.getAttributeValue("mail");
                            System.out.println("success");
                            System.out.println(dn);
                            System.out.println(name);
                            System.out.println(mail);
                        }
                    }
                    catch (LDAPSearchException lse)
                    {
                        // The search failed for some reason.
                        searchResult = lse.getSearchResult();
                        resultCode = lse.getResultCode();
                        String errorMessageFromServer = lse.getDiagnosticMessage();
                    }

                } catch (LDAPException e) {
                    login_flag=false;
                    e.printStackTrace();
                    //Toast.makeText(getBaseContext(),"No connection was established" , Toast.LENGTH_LONG).show();
                    System.out.println("No connection was established");
                } catch(Exception e) {
                    e.printStackTrace();
                } finally{
                    if(login_flag){
                        final SSLUtil sslUtil = new SSLUtil(new TrustAllTrustManager());
                        try {
                            LDAPConnection ldapConnection = new LDAPConnection(sslUtil.createSSLSocketFactory(),hostname,port);
                            ldapConnection.close();
                        } catch (LDAPException e) {
                            e.printStackTrace();
                        } catch (GeneralSecurityException e) {
                            e.printStackTrace();
                        }
                        //Toast.makeText(getBaseContext(), "Connection Closed successfully", Toast.LENGTH_LONG).show();
                        System.out.println("Connection Closed successfully");

                    }
                }
            }
        }).start();
    }

**Comment by dirmgr** [comment]
Your question sounds like it's more specific to the way that Active Directory works rather than an issue with the LDAP SDK. Active Directory sometimes does things in a way that's different from other servers, and I'm not as familiar with it. I think that the memberOf attribute comes into play for group membership, but I don't know the specifics.

**Comment by nikolai-krasilnikov** [comment]
Niel, hello. Thank you for your help. I found your answer on sourceforge and it helped me - https://sourceforge.net/p/ldap-sdk/discussion/1001257/thread/4ba7d9a1/#1829.

As a result, I implemented authorization with the following sequence of checks:
1) Using a service account, I check if there is such a user in AD and get his DN
2) Next, using CompareResult and the memberOf attribute, I check the membership of the security group.
3) I connect to the AD server using the user's DN and the entered password.
